Review enhancements

ayarotsky opened the pull request.

urkle committed:
Add base framework for process runner

- base gem setup with rspec, simplecov, and rubocop
- basic process management and registration with redis

urkle committed:
fixup! Add base framework for process runner

- base gem setup with rspec, simplecov, and rubocop
- basic process management and registration with redis

ayarotsky committed:
Review enhancements

urkle commented on lib/process_runner/cli.rb:
```diff
@@ -118,23 +117,16 @@ def parse_options(argv)
-        unless File.exist?(opts[:config_file])
-          raise ArgumentError, "Config file not found: #{opts[:config_file]}"
-        end
+      if opts[:config_file] && !File.exist?(opts[:config_file])
```
These are actually not the same thing.   As with your change here that means if I have a config file specified AND it exists, then it goes into the "else" condition, which it should not do.

urkle commented on lib/process_runner/cli.rb:
```diff
@@ -118,23 +117,16 @@ def parse_options(argv)
-            opts[:config_file] ||= path
-          end
-        end
+        path = File.join(config_dir, 'process_runner.yml')
```
so I had it the other way as I'm still debating on do I have it scan for process_runner.yml.erb (and add in ERB processing)..  (this is borrored from the sidekiq which checks for .yml and .yml.erb)

urkle commented on lib/process_runner/cli.rb:
```diff
@@ -234,6 +226,7 @@ def handle_signal(sig)
       end
     end
 
+    # Do we really need a wrapper around IO.select? It is used only in one place.
```
So this was to avoid some issues I was having in tests when stubbing IO.select.  I can add a comment in the method (or re-check to see if I can do things w/o the abstraction)

urkle commented on lib/process_runner/redis_connection.rb:
```diff
@@ -5,21 +5,20 @@
-      ConnectionPool.new(timeout: pool_timeout, size: size) do
-        build_client(options)
+    class << self
+      def create(options = {})
```
I actually prefer publicly expose class methods as self. and private ones in `class << self` as it makes it much clearer what is the public contract.

urkle commented on lib/process_runner/worker.rb:
```diff
@@ -19,6 +19,8 @@ def running?
     end
 
+    # Should `stopped?` be the complete opposite of `running?`?
+    # Should we also check `@origin` here?
```
no.. as it's not the same.    stopped? means it is fully stopped.   running? means it is actively running and not canceled.    thus `running?` and `stopped?` could both be false, in which case it is `stopping?`

urkle commented on spec/lib/process_runner/base_spec.rb:
```diff
@@ -8,7 +8,7 @@
       def lock_records; end
 
-      def process_record(_record) end
+      def process_record(_record); end
```
this is a 'fight' between the auto layout in rubymine..  as rubymine changes this variant to be 2 lines, which then rubocop yells about.   But if I have it one line w/o the semicolon (which is valid ruby) both are happy.

urkle commented on lib/process_runner/redis_connection.rb:
```diff
@@ -5,21 +5,20 @@
       def determine_redis_provider
-        p = ENV['REDIS_PROVIDER']
-        if p && p =~ /:/
+        if ENV['REDIS_PROVIDER'] =~ /[^\w_\d]/
```
pulled this change in..

urkle commented on lib/process_runner/lock/simple_redis.rb:
```diff
@@ -2,7 +2,7 @@
 module ProcessRunner
   module Lock
-    # this is a simple implementation of a lock t ensure only one job runner is running for a worker
+    # This is a simple implementation of a lock to ensure only one job runner is running for a worker
```
pulling this in.

urkle commented on lib/process_runner/base.rb:
```diff
@@ -14,10 +12,7 @@ def initialize(worker_index, options = {})
-        records&.each do |r|
-          process_record(r)
-        end
+        lock_records&.each(&method(:process_record))
```
I find this harder to read.

urkle committed:
fixup! Add base framework for process runner

- adjust worker_lock contract so locks can be extended
- add more logging
- add more thread names

urkle committed:
fixup! Add base framework for process runner

- add with advisory lock locking implementation
- remove no longer needed select proxy

urkle committed:
fixup! Add base framework for process runner

- add in proper method for configuring redis connection params (so it is the same server & client)

urkle committed:
fixup! Add base framework for process runner

- run check_worker when updating worker config

urkle committed:
fixup! Add base framework for process runner

- implement proper shutdown (stop workers + wait)

urkle committed:
fixup! Add base framework for process runner

- define a option controlled shutdown timeout

ayarotsky commented:
@urkle Out of curiosity, why did you decide to put `process_runner` under `exe` instead of `bin`, which is a standard folder for binary files in gems?

ayarotsky commented:
I think it's worth adding an example of the config file to the readme. It may contain all keys and their explanations/use cases.

ayarotsky committed:
fixup! Review enhancements

urkle committed:
fixup! Add base framework for process runner

urkle committed:
fixup! Add base framework for process runner

- add in releasing of Active record connections in rails mode

urkle committed:
fixup! Add base framework for process runner

- add stats collecting and reporting of running jobs
- add before and after perform hooks
- correct regex for redis ENV

urkle committed:
fixup! Add base framework for process runner

- implement running_workers and rename scheduled_workers (worker_count)

urkle committed:
fixup! Add base framework for process runner

- rubocop cleanup

urkle committed:
fixup! Add base framework for process runner

- expose out rejection reason for when worker fails

urkle committed:
fixup! Add base framework for process runner

- fix reloader handling in rails

urkle committed:
TEMP bump beta version

urkle committed:
fixup! Add base framework for process runner

- fix rubocop

urkle committed:
fixup! Add base framework for process runner

- fix rubocop

urkle committed:
rename gem (process_runner already taken)

urkle committed:
add stub loader
